File: PrototypeInterpreter.hs
{-# LANGUAGE NoOverloadedStrings #-}

module Main where

import Data.Char (isAlpha, isAlphaNum, isDigit)
import qualified Data.Map as M
import System.IO

-- ================================
-- TOKENS
-- ================================
data Token
    = NumberTok Double
    | IdentTok String
    | OpTok String
    | TAnd
    | TOr
    | TNot
    | TAssign
    | TLParen
    | TRParen
    | TSemicol
    | TEOF
    deriving (Show, Eq)


tokenize :: String -> [Token]
tokenize [] = [TEOF]
tokenize (c:cs)
    | c `elem` " \t" = tokenize cs
    | isDigit c =
        let (num, rest) = span (\x -> isDigit x || x == '.') (c:cs)
        in NumberTok (read num) : tokenize rest
    | isAlpha c =
        let (idstr, rest) = span isAlphaNum (c:cs)
        in case idstr of
            "and" -> TAnd : tokenize rest
            "or"  -> TOr  : tokenize rest
            "not" -> TNot : tokenize rest
            _     -> IdentTok idstr : tokenize rest
tokenize ('=':'=':cs) = OpTok "==" : tokenize cs
tokenize ('!':'=':cs) = OpTok "!=" : tokenize cs
tokenize ('>':'=':cs) = OpTok ">=" : tokenize cs
tokenize ('<':'=':cs) = OpTok "<=" : tokenize cs
tokenize (c:cs)
    | c `elem` "+-*/%><" = OpTok [c] : tokenize cs
    | c == '=' = TAssign : tokenize cs
    | c == '(' = TLParen : tokenize cs
    | c == ')' = TRParen : tokenize cs
    | c == ';' = TSemicol : tokenize cs
    | otherwise = error ("Unexpected char: " ++ [c])


-- ================================
-- AST
-- ================================
data AST
    = NumLit Double
    | Var String
    | Assign String AST
    | Unary String AST
    | BinOp String AST AST
    deriving (Show, Eq)

-- ================================
-- PARSER
-- ================================
newtype ParserState = PS { ts :: [Token] }

peek :: ParserState -> Token
peek (PS (t:_)) = t
peek _ = TEOF

advance :: ParserState -> (Token, ParserState)
advance (PS (t:rest)) = (t, PS rest)
advance s = (TEOF, s)

type Parser a = ParserState -> (a, ParserState)

expect :: (Token -> Bool) -> Parser Token
expect cond (PS (t:rest))
    | cond t = (t, PS rest)
    | otherwise = error ("Unexpected token: " ++ show t)
expect _ _ = error "Unexpected EOF"

-- Grammar functions

parseProgram :: Parser [AST]
parseProgram st =
    let (stmt1, st1) = parseStmt st
    in case peek st1 of
        TSemicol ->
            let (_, st2) = advance st1
                (rest, st3) = parseProgram st2
            in (stmt1 : rest, st3)
        _ -> ([stmt1], st1)

parseStmt :: Parser AST
parseStmt st =
    case peek st of
        IdentTok _ ->
            case ts st of
                (IdentTok v : TAssign : _) ->
                    let (_, st1) = advance st
                        (_, st2) = advance st1
                        (expr, st3) = parseExpr st2
                    in (Assign v expr, st3)
                _ -> parseExpr st
        _ -> parseExpr st

parseExpr :: Parser AST
parseExpr = parseOr

parseOr :: Parser AST
parseOr st =
    let (left, st1) = parseAnd st
    in case peek st1 of
        TOr ->
            let (_, st2) = advance st1
                (right, st3) = parseAnd st2
            in (BinOp "or" left right, st3)
        _ -> (left, st1)

parseAnd :: Parser AST
parseAnd st =
    let (left, st1) = parseEquality st
    in case peek st1 of
        TAnd ->
            let (_, st2) = advance st1
                (right, st3) = parseEquality st2
            in (BinOp "and" left right, st3)
        _ -> (left, st1)

parseEquality :: Parser AST
parseEquality st =
    let (left, st1) = parseComparison st
    in case peek st1 of
        OpTok "==" ->
            let (_, st2) = advance st1
                (right, st3) = parseComparison st2
            in (BinOp "==" left right, st3)
        OpTok "!=" ->
            let (_, st2) = advance st1
                (right, st3) = parseComparison st2
            in (BinOp "!=" left right, st3)
        _ -> (left, st1)

parseComparison :: Parser AST
parseComparison st =
    let (left, st1) = parseTerm st
    in case peek st1 of
        OpTok o | o `elem` [">","<",">=","<="] ->
            let (_, st2) = advance st1
                (right, st3) = parseTerm st2
            in (BinOp o left right, st3)
        _ -> (left, st1)

parseTerm :: Parser AST
parseTerm st =
    let (left, st1) = parseFactor st
    in case peek st1 of
        OpTok o | o `elem` ["+","-"] ->
            let (_, st2) = advance st1
                (right, st3) = parseFactor st2
            in (BinOp o left right, st3)
        _ -> (left, st1)

parseFactor :: Parser AST
parseFactor st =
    let (left, st1) = parseUnary st
    in case peek st1 of
        OpTok o | o `elem` ["*","/","%"] ->
            let (_, st2) = advance st1
                (right, st3) = parseUnary st2
            in (BinOp o left right, st3)
        _ -> (left, st1)

parseUnary :: Parser AST
parseUnary st =
    case peek st of
        TNot ->
            let (_, st1) = advance st
                (expr, st2) = parseUnary st1
            in (Unary "not" expr, st2)
        _ -> parsePrimary st

parsePrimary :: Parser AST
parsePrimary st =
    case peek st of
        NumberTok n ->
            let (_, st1) = advance st
            in (NumLit n, st1)

        IdentTok v ->
            let (_, st1) = advance st
            in (Var v, st1)

        TLParen ->
            let (_, st1) = advance st
                (expr, st2) = parseExpr st1
            in case peek st2 of
                TRParen ->
                    let (_, st3) = advance st2
                    in (expr, st3)
                _ -> error "Missing )"

        t -> error ("Unexpected token: " ++ show t)

-- ================================
-- INTERPRETER
-- ================================
type Env = M.Map String Double

evalAST :: Env -> AST -> (Double, Env)
evalAST env (NumLit n) = (n, env)

evalAST env (Var v) =
    case M.lookup v env of
        Just val -> (val, env)
        Nothing -> error ("Undefined variable: " ++ v)

evalAST env (Assign v expr) =
    let (val, env2) = evalAST env expr
        env3 = M.insert v val env2
    in (val, env3)

evalAST env (Unary "not" expr) =
    let (v, env2) = evalAST env expr
        result = if v == 0 then 1 else 0
    in (result, env2)

evalAST env (BinOp op l r) =
    let (lv, env1) = evalAST env l
        (rv, env2) = evalAST env1 r
        result = case op of
            "+" -> lv + rv
            "-" -> lv - rv
            "*" -> lv * rv
            "/" -> lv / rv
            "%" -> fromIntegral (floor lv `mod` floor rv)
            "==" -> boolToNum (lv == rv)
            "!=" -> boolToNum (lv /= rv)
            "<"  -> boolToNum (lv < rv)
            ">"  -> boolToNum (lv > rv)
            "<=" -> boolToNum (lv <= rv)
            ">=" -> boolToNum (lv >= rv)
            "and" -> boolToNum ((lv /= 0) && (rv /= 0))
            "or"  -> boolToNum ((lv /= 0) || (rv /= 0))
            _ -> error ("Unknown operator: " ++ op)
    in (result, env2)

boolToNum :: Bool -> Double
boolToNum b = if b then 1 else 0

-- ================================
-- REPL
-- ================================
main :: IO ()
main = do
    putStrLn "=== Advanced Expression Interpreter (Haskell) ==="
    putStrLn "Type expressions or assignments. Use ';' to chain statements."
    putStrLn "Type 'exit' to quit.\n"
    loop M.empty
  where
    loop env = do
        putStr ">>> "
        hFlush stdout
        line <- getLine
        if line == "exit"
            then return ()
            else do
                let toks = tokenize line
                    (stmts, _) = parseProgram (PS toks)
                    (result, newEnv) = evalMany env stmts
                print result
                loop newEnv

    evalMany :: Env -> [AST] -> (Double, Env)
    evalMany env [] = (0, env)
    evalMany env [x] = evalAST env x
    evalMany env (x:xs) =
        let (_, env2) = evalAST env x
        in evalMany env2 xs

File: PrototypeInterpreter.py
import re

# ================================
# LEXER
# ================================
TOKEN_SPEC = [
    ("NUMBER",   r'\d+(\.\d+)?'),
    ("ID",       r'[A-Za-z_]\w*'),
    ("OP",       r'==|!=|>=|<=|>|<|\+|\-|\*|/|%'),
    ("AND",      r'and'),
    ("OR",       r'or'),
    ("NOT",      r'not'),
    ("ASSIGN",   r'='),
    ("LPAREN",   r'\('),
    ("RPAREN",   r'\)'),
    ("SEMICOL",  r';'),
    ("SKIP",     r'[ \t]+'),
    ("MISMATCH", r'.')
]

TOKEN_REGEX = "|".join(f"(?P<{name}>{pattern})" for name, pattern in TOKEN_SPEC)

def tokenize(code):
    tokens = []
    for match in re.finditer(TOKEN_REGEX, code):
        kind = match.lastgroup
        value = match.group()

        if kind == "NUMBER":
            value = float(value) if '.' in value else int(value)
        elif kind == "ID":
            pass
        elif kind == "SKIP":
            continue
        elif kind == "MISMATCH":
            raise SyntaxError(f"Unexpected token: {value}")

        tokens.append((kind, value))
    tokens.append(("EOF", None))
    return tokens

# ================================
# PARSER (RECURSIVE DESCENT)
# ================================
class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos]

    def advance(self):
        self.pos += 1
        return self.tokens[self.pos - 1]

    # Grammar:
    # program     → stmt (';' stmt)*
    # stmt        → ID '=' expr | expr
    # expr        → logic_or
    # logic_or    → logic_and ('or' logic_and)*
    # logic_and   → equality ('and' equality)*
    # equality    → comparison ( (== | != ) comparison )*
    # comparison  → term ( > | < | >= | <= term )*
    # term        → factor ( ('+'|'-') factor )*
    # factor      → unary ( ('*'|'/'|'%') unary )*
    # unary       → 'not' unary | primary
    # primary     → NUMBER | ID | '(' expr ')'

    def parse(self):
        statements = [self.statement()]
        while self.peek()[0] == "SEMICOL":
            self.advance()
            statements.append(self.statement())
        return statements

    def statement(self):
        tok_type, tok_val = self.peek()
        if tok_type == "ID":
            # Lookahead for assignment
            next_tok = self.tokens[self.pos + 1]
            if next_tok[0] == "ASSIGN":
                var = self.advance()[1]    # ID
                self.advance()             # =
                expr = self.expr()
                return ("assign", var, expr)
        return self.expr()

    def expr(self):
        return self.logic_or()

    def logic_or(self):
        node = self.logic_and()
        while self.peek()[0] == "OR":
            op = self.advance()[1]
            right = self.logic_and()
            node = ("binop", op, node, right)
        return node

    def logic_and(self):
        node = self.equality()
        while self.peek()[0] == "AND":
            op = self.advance()[1]
            right = self.equality()
            node = ("binop", op, node, right)
        return node

    def equality(self):
        node = self.comparison()
        while self.peek()[0] == "OP" and self.peek()[1] in ("==", "!="):
            op = self.advance()[1]
            right = self.comparison()
            node = ("binop", op, node, right)
        return node

    def comparison(self):
        node = self.term()
        while self.peek()[0] == "OP" and self.peek()[1] in (">", "<", ">=", "<="):
            op = self.advance()[1]
            right = self.term()
            node = ("binop", op, node, right)
        return node

    def term(self):
        node = self.factor()
        while self.peek()[0] == "OP" and self.peek()[1] in ("+", "-"):
            op = self.advance()[1]
            right = self.factor()
            node = ("binop", op, node, right)
        return node

    def factor(self):
        node = self.unary()
        while self.peek()[0] == "OP" and self.peek()[1] in ("*", "/", "%"):
            op = self.advance()[1]
            right = self.unary()
            node = ("binop", op, node, right)
        return node

    def unary(self):
        tok_type, tok_val = self.peek()
        if tok_type == "NOT":
            self.advance()
            return ("unop", "not", self.unary())
        return self.primary()

    def primary(self):
        tok_type, tok_val = self.peek()

        if tok_type == "NUMBER":
            self.advance()
            return ("num", tok_val)

        if tok_type == "ID":
            self.advance()
            return ("var", tok_val)

        if tok_type == "LPAREN":
            self.advance()
            expr = self.expr()
            if self.peek()[0] != "RPAREN":
                raise SyntaxError("Missing closing parenthesis")
            self.advance()
            return expr

        raise SyntaxError(f"Unexpected token: {tok_val}")

# ================================
# INTERPRETER (EVALUATOR)
# ================================
class Interpreter:
    def __init__(self):
        self.env = {}  # symbol table

    def eval(self, node):
        ntype = node[0]

        if ntype == "num":
            return node[1]

        if ntype == "var":
            name = node[1]
            if name not in self.env:
                raise NameError(f"Undefined variable '{name}'")
            return self.env[name]

        if ntype == "assign":
            _, var, expr = node
            value = self.eval(expr)
            self.env[var] = value
            return value

        if ntype == "unop":
            _, op, expr = node
            val = self.eval(expr)
            if op == "not":
                return not val

        if ntype == "binop":
            _, op, left, right = node
            l = self.eval(left)
            r = self.eval(right)

            return {
                "+": l + r,
                "-": l - r,
                "*": l * r,
                "/": l / r,
                "%": l % r,
                "==": l == r,
                "!=": l != r,
                ">": l > r,
                "<": l < r,
                ">=": l >= r,
                "<=": l <= r,
                "and": l and r,
                "or": l or r
            }[op]

        raise RuntimeError("Unknown AST node")

# ================================
# MAIN REPL LOOP
# ================================
print("=== Advanced Expression Interpreter ===")
print("Type expressions or assignments. Use ';' to chain statements.")
print("Type 'exit' to quit.\n")

interpreter = Interpreter()

while True:
    code = input(">>> ")

    if code.strip() == "exit":
        break

    try:
        tokens = tokenize(code)
        parser = Parser(tokens)
        statements = parser.parse()

        last_value = None
        for stmt in statements:
            last_value = interpreter.eval(stmt)

        print(last_value)

    except Exception as e:
        print("Error:", e)
